### ISSUE TITLE
parent based sampling not working?

### ISSUE THREAD

**Issue by kubav182** [microsoft/ApplicationInsights-Java#3955]
Hello, I'm facing issue and need some help.

I have service A with sampling overrides dependency for 0 %. 
I want to turn off logging just for the dependency, I don't want tell what should happen inside the dependency itself. Dependency log is missing so it works as expected.

Service A calls service B. In service B I receive traceparent header with suffix -00 even there was -01 on service A. So sampling dependency changes this flag and tells what should happen inside dependency B. Thats something I don't want.
Service B logs request, but I'm missing all logs after the request log like traces and dependencies. Request log has sampling attribute 1. It does not affect b3 headers only traceparent header.

I found undocumented property preview->sampling->parentBased, but setting this to false has no effect same as setting this to true.

I'm using Spring Boot 3.2 with actuator, otel micrometer and insights 3.6.2. I also tried to change opentelemetry sampler bean so it is not parent based.

Can you please guide me how can I achieve what I want so i will not log dependency on service A, but inside dependency B I will see logs?

The reason is reducing cost as we don't need to log dependencies to internal services we are ok with request logs so we log dependencies only to external services.

Thanks

**Comment by jeanbisutti** [comment]
Hi @kubav182 

Do you mean that:
1)  You have an HTTP endpoint A calling an HTTP endpoint point B
2) You want to have the dependency telemetry, for example a database call, for the HTTP endpoint B but not for the HTTP endpoint A

What do you mean by "external sevice"? Is it a service provided by another Spring Boot application and instrumented by the Application Insights Java agent? Is service B an "external service"?

Thank you

**Comment by kubav182** [comment]
Hello @jeanbisutti maybe I wrote it too complicated. I want to see all logs on service B, but don't want to see dependency log on service A which is calling service B. Thats all. Service A and B are internal microservices using same stack.

Service A has applicationinsights.json config it overrides sampling to 0% for dependency log and by regexp is filtered service B. I really can't see dependency logs just for service B its ok. But it causes service B is not logging trace logs. Reason is that traceparent header is modified by service A so it ends with -00 flag instead of -01.

**Comment by jeanbisutti** [comment]
@kubav182 You could disable some instrumentations in service A to reduce cost: https://learn.microsoft.com/en-us/azure/azure-monitor/app/java-standalone-config#suppress-specific-autocollected-telemetry

**Comment by kubav182** [comment]
@jeanbisutti We already do this. But we did not have this problem with SB2 and b3 headers. We have this problem with SB3 and traceparent header. So now logging dependencies that we did not log in the past means +16% in cost for logs we dont need and dont use. If we turn on sampling for these dependencies we miss other critical logs.

I was hoping there is a solution to ommit this behaviour. For example config preview->sampling->parentBased does not work for me.

**Comment by jeanbisutti** [comment]
@kubav182  

>So now logging dependencies that we did not log in the past means +16% in cost for logs we dont need and dont use. 

Could you please provide information about the unwanted dependencies (database dependency, other)?

**Comment by kubav182** [comment]
Service A receives HTTP call with traceparent header 00-80e1afed08e019fc1110464cfa66635c-7a085853722dc6d2-01
Service A overrides sampling to 0% for dependency logs when calling service B via HTTP.
Service A calls service B via HTTP with traceparent header 00-80e1afed08e019fc1110464cfa66635c-7a085853722dc6d2-00
Service B does not log traces because of -00 suffix in traceparent header.

**Comment by jeanbisutti** [comment]
> [@jeanbisutti](https://github.com/jeanbisutti) it is HTTP call to service B with feign client

Feign does not seem instrumented by the OpenTelemetry Java agent (https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/3606, https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/5678).


> +16% in cost for logs we dont need and dont use

If by log you mean telemetry data, you could add sampling overrides with low sampling values.

**Comment by kubav182** [comment]
@jeanbisutti We have sampling configuration which follows our needs, we don't want to change it. It worked for many years. Problem is Spring Boot 3 + traceparent header + insights agent 3.x as it does something undocumented and the solution is also undocumented so we just want to achieve what we did in previous version.

When service samples dependency logs via json config it changes http traceparent header. It does not affect b3 headers and probably never did, so it may be the reason why it worked in previous versions with sleuth and b3 headers.
So we want to achieve one of this options:
A. Dont change output traceparent header suffix to -00
B. If receive suffix -00 as input header ignore it and behave as it was -01. Not sure why undocumented config preview->sampling->parentBased has no effect.

**Comment by jeanbisutti** [comment]
@kubav182 

It would help if you could provide a minimal Github project showing how Service A is calling Service B.

Have you observed the issue by migrating from Spring Boot 2 to Spring Boot 3?

**Comment by kubav182** [comment]
@jeanbisutti yes I observed it during update from SB2 to SB3, it was harder to observe as telemetry of service B is now affected by configuration on service A and the update was not in same time. Now I discovered the problem appears when these dependencies are on classpath at the same time. (We dont have insights on all environments)
implementation("com.microsoft.azure:applicationinsights-runtime-attach:3.6.2")
implementation("io.micrometer:micrometer-tracing-bridge-otel")
implementation("io.github.openfeign:feign-micrometer")

If I comment micrometer dependencies I can see traceparent header suffix is modified by agent to -00 but service B is still logging.
If i add micrometer then service B is not logging (except request log) when traceparent header suffix is modified to -00 and it is modified.
If I have micrometer still on classpath and use insights config preview -> disablePropagation then traceparent header is not modifed so service B is logging.

Is preview -> disablePropagation expected solution to solve this usecase?
I can provide simple app if you still want.
